passes/cmds/partition.cc

```diff
@@ -0,0 +1,501 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2020  Alberto Gonzalez <boqwxp@airmail.cc>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/cost.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct PartitionPass : public Pass {
+	//Constructs a hypergraph from a flat techmapped module, partitions the hypergraph,
+	//recovers the cell partition assignments, creates partition modules with the appropriate
+	//cells, ports, and intra-partition connections, then replaces the original module
+	//with one that instantiates and connects the partition modules.
+	int opt_k, opt_s, opt_imbalance;
+	bool opt_verbose, opt_nocleanup;
+	RTLIL::Module *module, *new_module;
+	RTLIL::Design *design;
+	dict<RTLIL::IdString, int> cell_type_costs;
+	dict<RTLIL::SigBit, int> sigbit_to_edgenum;
+	dict<int, pool<int>> edge_nodes;
+	dict<int, pool<int>> partition_nodes;
+	dict<int, bool> edge_cut;
+	pool<RTLIL::IdString> cut_wires;
+	int node_ctr, edge_ctr, cut_edge_ctr, partitions;
+	std::vector<RTLIL::Module *> partition_modules;
+
+	PartitionPass() : Pass("partition", "split techmapped module into equal-sized parts"), opt_k(-1), opt_s(-1),
```

Comment:
Can you use C++11 member initializers (I think that's what they're called) instead of these gigantic lists that (a) produce warnings if options are out of order and (b) make it easy to omit a member leading to UB?

```
RTLIL::Design *design = nullptr;
```

Reply:
You mean [initializer_list](https://en.cppreference.com/w/cpp/utility/initializer_list)?  Not sure I really see the benefit here.  Can you ELI5?

Reply:
I mean [in-class member initializers](https://www.stroustrup.com/C++11FAQ.html#member-init).

Reply:
Oh shit, how have I missed that C++11 feature?! Yeah, that's way better.

Reply:
Also if you're feeling like doing it, I'd be on board with doing this conversion for the rest of Yosys.

Reply:
Maybe some day.  But not today :P.